gitfile/tokenfetcher_spi.go

```diff
@@ -0,0 +1,181 @@
+	namespace, err := ioutil.ReadFile(namespaceFile)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		panic(err.Error())
+	}
+
+	scheme := runtime.NewScheme()
+	if err = corev1.AddToScheme(scheme); err != nil {
+		panic(err.Error())
+	}
+
+	if err = v1beta1.AddToScheme(scheme); err != nil {
+		panic(err.Error())
+	}
+
+	// creates the client
+	k8sClient, err := client.New(config, client.Options{Scheme: scheme})
+	if err != nil {
+		panic(err.Error())
+	}
+	return &SpiTokenFetcher{k8sClient: k8sClient, namespace: string(namespace)}
+}
+
+func (s *SpiTokenFetcher) BuildHeader(ctx context.Context, repoUrl string, loginCallback func(url string)) (*HeaderStruct, error) {
+
+	var tBindingName = "file-retriever-binging-" + randStringBytes(6)
+	var secretName = "file-retriever-secret-" + randStringBytes(5)
+
+	// create binding
+	newBinding := newSPIATB(tBindingName, s, repoUrl, secretName)
+	err := s.k8sClient.Create(ctx, newBinding)
+	if err != nil {
+		zap.L().Error("Error creating Token Binding item:", zap.Error(err))
+		return nil, err
+	}
+
+	// scheduling the binding cleanup
+	defer func() {
+		// clean up token binding
+		err = s.k8sClient.Delete(ctx, newBinding)
+		if err != nil {
+		}
+	}()
+
+	// now re-reading SPITokenBinding to get updated fields
+	var tokenName string
+	for {
+		readBinding := &v1beta1.SPIAccessTokenBinding{}
+		err = s.k8sClient.Get(ctx, client.ObjectKey{Namespace: s.namespace, Name: tBindingName}, readBinding)
+		if err != nil {
+			zap.L().Error("Error reading TB item:", zap.Error(err))
+			return nil, err
+		}
+		tokenName = readBinding.Status.LinkedAccessTokenName
+		if tokenName != "" {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("task is cancelled")
+		default:
+			time.Sleep(200 * time.Millisecond)
+		}
+	}
+	zap.L().Info(fmt.Sprintf("Access Token to watch: %s", tokenName))
+
+	// now try read SPIAccessToken to get link
+	var url string
+	var loginCalled = false
+	for {
+		readToken := &v1beta1.SPIAccessToken{}
+		_ = s.k8sClient.Get(ctx, client.ObjectKey{Namespace: s.namespace, Name: tokenName}, readToken)
+		if readToken.Status.Phase == v1beta1.SPIAccessTokenPhaseAwaitingTokenData && !loginCalled {
+			url = readToken.Status.OAuthUrl
+			zap.L().Info(fmt.Sprintf("URL to OAUth: %s", url))
+			go loginCallback(url)
+			loginCalled = true
+		} else if readToken.Status.Phase == v1beta1.SPIAccessTokenPhaseReady {
+			// now we can exit the loop and read the secret
+			break
+		}
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("task is cancelled")
+		default:
+			time.Sleep(200 * time.Millisecond)
+		}
+	}
+
+	// reading token secret
+	tokenSecret := &corev1.Secret{}
+	err = s.k8sClient.Get(ctx, client.ObjectKey{Namespace: s.namespace, Name: secretName}, tokenSecret)
+	if err != nil {
+		zap.L().Error("Error reading Token Secret item:", zap.Error(err))
+		return nil, err
+	}
```

Comment:
This should be in a loop, because the creation of the secret might not be immediate..

Reply:
isn't `SPIAccessTokenPhaseReady` phase means secret is created ?